src/userobjects/OpenMCDomainFilterEditor.C

```diff
@@ -0,0 +1,202 @@
+  return params;
+}
+
+OpenMCDomainFilterEditor::OpenMCDomainFilterEditor(const InputParameters & parameters)
+  : OpenMCUserObject(parameters),
+    _create_filter(getParam<bool>("create_filter")),
+    _filter_id(getParam<int32_t>("filter_id")),
+    _filter_type(getParam<MooseEnum>("filter_type").getEnum<OpenMCFilterType>())
+{
+
+  if (!parameters.isParamSetByUser("filter_type"))
+  {
+    mooseError(longName() + ": filter_type must be specified");
+  }
+
+  this->initialize();
+}
+
+void
+OpenMCDomainFilterEditor::initialize()
+{
+  bool filter_exists = this->filterExists();
+
+  // if create_filter is set to true, but the filter already exists, display a warning
+  if (_create_filter && filter_exists)
+  {
+    mooseWarning(longName() + ": Filter " + std::to_string(_filter_id) +
+                 " already exists in the OpenMC XML model");
+  }
+
+  if (!_create_filter && !filter_exists)
+    paramError("filter_id", "Filter " + std::to_string(_filter_id) + " does not exist and create_filter is false");
+
+  // if the filter doesn't exist at this point and no other errors have been raised,
+  // create the filter
+  if (!filter_exists)
+    openmc::Filter::create(filterTypeEnumToString(_filter_type), _filter_id);
+
+  // at this point the filter exists and the filter type is set, check that the type is valid
+  checkFilterTypeMatch();
+}
+
+void
+OpenMCDomainFilterEditor::checkFilterTypeMatch() const
+{
+  // check this UO's filter type against the one in the OpenMC model
+  std::string existing_type_str = openmc::model::tally_filters[filterIndex()]->type_str();
+  OpenMCFilterType existing_type = stringToFilterTypeEnum(existing_type_str);
+  _console << "CHECKING FILTER TYPE..." << std::endl;
+
+  if (existing_type != _filter_type)
+    paramError("filter_id", "An existing filter, Filter " + std::to_string(_filter_id) + ", is of type \"" + existing_type_str + "\" and cannot be changed to type \"" + filterTypeEnumToString(_filter_type) + "\"");
+}
+
+bool
+OpenMCDomainFilterEditor::filterExists() const
+{
+  return openmc::model::filter_map.find(_filter_id) != openmc::model::filter_map.end();
+}
+
+int32_t
+OpenMCDomainFilterEditor::filterIndex() const
+{
+  if (!filterExists())
+    mooseError(longName() + ": Filter " + std::to_string(_filter_id) + " does not exist");
+  return openmc::model::filter_map.at(_filter_id);
+}
+
+void
+OpenMCDomainFilterEditor::execute()
+{
+  openmc::Filter * filter = openmc::model::tally_filters[filterIndex()].get();
+
+  // TODO: update if webcontrols starts to support integral types
+  std::vector<int32_t> ids, bins;
+  for (auto bin_id : getParam<std::vector<std::string>>("bins"))
+  {
+    ids.push_back(std::stoi(bin_id));
+  }
+
+  if (_filter_type == OpenMCFilterType::cell)
+  {
+    openmc::CellFilter * cell_filter = dynamic_cast<openmc::CellFilter *>(filter);
+    if (!cell_filter)
+      mooseError(longName() + ": Filter " + std::to_string(_filter_id) + " is not a cell filter");
```

Comment:
Same here, for the cell, material, universe, and mesh filter checks -- I don't think there are tests covering these

Reply:
I'm honestly not sure how I'd trigger these errors from the test system, which might indicate that we don't need them, but with `dynamic_cast`'s going on here I think it's wise to keep them so we get some kind of output instead of a segfault if something goes wrong.

Reply:
Sounds good, that's fine by me!